src/chttpd/src/chttpd_misc.erl

```diff
@@ -146,6 +147,143 @@ handle_all_dbs_req(#httpd{method='GET'}=Req) ->
+            end, "", Result),
+            send_chunk(Resp, "]"),
+            chttpd:end_json_response(Resp);
+        Error ->
+            throw(Error)
+    end;
+deleted_dbs_info_req(Req, _DbName) ->
+    send_method_not_allowed(Req, "GET,HEAD").
+
+deleted_dbs_post_req(#httpd{user_ctx=Ctx}=Req) ->
+    couch_httpd:verify_is_server_admin(Req),
+    chttpd:validate_ctype(Req, "application/json"),
+    {JsonProps} = chttpd:json_body_obj(Req),
+    UndeleteJson0 = couch_util:get_value(<<"undelete">>, JsonProps, undefined),
+    DeleteJson0 = couch_util:get_value(<<"delete">>, JsonProps, undefined),
+    case {UndeleteJson0, DeleteJson0}  of
+        {undefined, undefined} ->
+            throw({bad_request, <<"POST body must include `undeleted` "
+                "or `delete` parameter.">>});
+        {UndeleteJson, undefined} ->
+            handle_undelete_db_req(Req, UndeleteJson);
+        {undefined, DeleteJson} ->
+            handle_delete_deleted_req(Req, DeleteJson);
+        {_Else, _Else} ->
+            throw({bad_request,
+                <<"`undeleted` and `delete` are mutually exclusive">>})
+    end.
+
+handle_undelete_db_req(#httpd{user_ctx=Ctx}=Req, {JsonProps}) ->
+    DbName = case couch_util:get_value(<<"source">>, JsonProps) of
+        undefined ->
+            throw({bad_request,
+                <<"POST body must include `source` parameter.">>});
+        DbName0 ->
+            DbName0
+    end,
+    TimeStamp = case couch_util:get_value(<<"source_timestamp">>, JsonProps) of
+        undefined ->
+            throw({bad_request,
+                <<"POST body must include `source_timestamp` parameter.">>});
+        TimeStamp0 ->
+            TimeStamp0
+    end,
+    TgtDbName = case couch_util:get_value(<<"target">>, JsonProps) of
+        undefined ->  DbName;
+        TgtDbName0 -> TgtDbName0
+    end,
+    case fabric2_db:undelete(DbName, TgtDbName, TimeStamp, [{user_ctx, Ctx}]) of
+        ok ->
+            send_json(Req, 200, {[{ok, true}]});
+        {error, file_exists} ->
+            chttpd:send_error(Req, file_exists);
+        Error ->
+            throw(Error)
+    end.
+
+handle_delete_deleted_req(#httpd{user_ctx=Ctx}=Req, {JsonProps}) ->
+    DbName = case couch_util:get_value(<<"source">>, JsonProps) of
+        undefined ->
+            throw({bad_request,
+                <<"POST body must include `source` parameter.">>});
+        DbName0 ->
+            DbName0
+    end,
+    TimeStamp = case couch_util:get_value(<<"source_timestamp">>, JsonProps) of
+        undefined ->
+            throw({bad_request,
+                <<"POST body must include `source_timestamp` parameter.">>});
+        TimeStamp0 ->
+            TimeStamp0
+    end,
+    case fabric2_db:delete_deleted(DbName, TimeStamp, [{user_ctx, Ctx}]) of
+        ok ->
+            send_json(Req, 200, {[{ok, true}]});
+        {error, file_exists} ->
+            chttpd:send_error(Req, file_exists);
```

Comment:
I don't think file_exists is the right error here. I should be not_found, shouldn't it?